dbms/src/Functions/FunctionsIsIPAddr.h

```diff
@@ -0,0 +1,288 @@
+            const typename ColumnString::Chars_t & data = col_input->getChars();
+            const typename ColumnString::Offsets & offsets = col_input->getOffsets();
+
+            auto col_res = ColumnUInt8::create();
+            ColumnUInt8::Container & vec_res = col_res->getData();
+            vec_res.resize(size);
+
+            size_t prev_offset = 0;
+            for (size_t i = 0; i < size; ++i)
+            {
+                vec_res[i] = static_cast<UInt8>(isIPv4(reinterpret_cast<const char *>(&data[prev_offset])));
+                prev_offset = offsets[i];
+            }
+
+            block.getByPosition(result).column = std::move(col_res);
+        }
+        else
+            throw Exception(
+                fmt::format("Illegal column {} of argument of function {}", block.getByPosition(arguments[0]).column->getName(), getName()),
+                ErrorCodes::ILLEGAL_COLUMN);
+    }
+};
+
+class FunctionIsIPv6 : public IFunction
+{
+public:
+    static constexpr auto name = "tiDBIsIPv6";
+    FunctionIsIPv6() = default;
+
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionIsIPv6>(); };
+
+    std::string getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 1)
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be 1.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        return std::make_shared<DataTypeUInt8>();
+    }
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        if (const auto * col_input = checkAndGetColumn<ColumnString>(block.getByPosition(arguments[0]).column.get()))
+        {
+            size_t size = block.getByPosition(arguments[0]).column->size();
+            const typename ColumnString::Chars_t & data = col_input->getChars();
+            const typename ColumnString::Offsets & offsets = col_input->getOffsets();
+
+            auto col_res = ColumnUInt8::create();
+            ColumnUInt8::Container & vec_res = col_res->getData();
+            vec_res.resize(size);
+
+            size_t prev_offset = 0;
+            for (size_t i = 0; i < size; ++i)
+            {
+                vec_res[i] = static_cast<UInt8>(isIPv6(reinterpret_cast<const char *>(&data[prev_offset])));
+                prev_offset = offsets[i];
+            }
+
+            block.getByPosition(result).column = std::move(col_res);
+        }
+        else
+            throw Exception(
+                fmt::format("Illegal column {} of argument of function {}", block.getByPosition(arguments[0]).column->getName(), getName()),
+                ErrorCodes::ILLEGAL_COLUMN);
+    }
+};
+} // namespace DB
```

Comment:
The implementation of `executeImpl` in ipv4 and ipv6 is similiar, and I think we can define a function to unify the logic such as
```
template <typename Impl>
void executeCheckIP(xxx)
{
...
  vec_res[i] = static_cast<UInt8>(Impl::check(reinterpret_cast<const char *>(&data[prev_offset])));
...
}

returnType executeImpl(xxx)
{
  return executeCheckIP(xxx);
}
```

Reply:
Thanks a lot, I referenced other functions. The following template is used
```C++
template <typename Name, UInt8(Function)(const char *)>
```
Is this okay?

Reply:
> Thanks a lot, I referenced other functions. The following template is used
> 
> ```c++
> template <typename Name, UInt8(Function)(const char *)>
> ```
> 
> Is this okay?

I think it's okk